### PR TITLE
Add Code Gen piece for Recommendation task

### PR DESCRIPTION
**Pull request by maryamariyan**
- [x] Adds CodeGen piece for recommendation task
- [x] Cherry-picks changes in https://github.com/dotnet/machinelearning/pull/4246 into features/automl branch
- [x] Cross check mlnet tests build and run fine
- [ ] confirm AutoML project tests build and run fine

cc: @LittleLittleCloud @eerhardt

### CODE REVIEW

File: src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj [hunk 1]
```diff
@@ -11,10 +11,13 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.ML" Version="$(MlDotNetPackageVersion)" />
-    <PackageReference Include="Microsoft.ML.LightGBM" Version="$(MlDotNetPackageVersion)" />
-    <PackageReference Include="Microsoft.ML.Mkl.Components" Version="$(MlDotNetPackageVersion)" />
-    <PackageReference Include="Microsoft.ML.Recommender" Version="$(MlDotNetPackageVersion)" />
+    <PackageReference Include="Microsoft.ML" Version="1.4.0-preview3-28218-2" />
+    <PackageReference Include="Microsoft.ML.LightGBM" Version="1.4.0-preview3-28218-2" />
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj]: Should it be `LightGbm`?

File: Microsoft.ML.AutoML.sln [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.28010.2050	
+# Visual Studio Version 16
```
Comment (LittleLittleCloud) [inline Microsoft.ML.AutoML.sln]: Is this change necessary?

File: src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj (hunk 1 above)
Comment (maryamariyan) [reply src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj]: note: LightGbm in all uppercasing is also in nupkgproj files (pkg\..AutoML and pkg\mlnet)

File: Microsoft.ML.AutoML.sln [hunk 3]
```diff
@@ -18,6 +18,8 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "tools-local", "tools-local"
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.ML.CodeGenerator", "src\Microsoft.ML.CodeGen\Microsoft.ML.CodeGenerator.csproj", "{3148A8CF-A686-4A9C-A5C6-C3E41CAAD64D}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.ML.Recommender", "src\Microsoft.ML.Recommender\Microsoft.ML.Recommender.csproj", "{19D6A195-657E-42D9-A5D4-ECB289BB6FDA}"
+EndProject
```
Comment (LittleLittleCloud) [inline Microsoft.ML.AutoML.sln]: Why `Recommender` appears in AutoML solution?

File: src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj [hunk 4]
```diff
@@ -11,9 +11,13 @@
...
-    <PackageReference Include="Microsoft.ML" Version="$(MlDotNetPackageVersion)" />
-    <PackageReference Include="Microsoft.ML.LightGBM" Version="$(MlDotNetPackageVersion)" />
-    <PackageReference Include="Microsoft.ML.Mkl.Components" Version="$(MlDotNetPackageVersion)" />
+    <PackageReference Include="Microsoft.ML" Version="1.4.0-preview3-28218-2" />
+    <PackageReference Include="Microsoft.ML.LightGBM" Version="1.4.0-preview3-28218-2" />
+    <PackageReference Include="Microsoft.ML.Mkl.Components" Version="1.4.0-preview3-28218-2" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Microsoft.ML.Recommender\Microsoft.ML.Recommender.csproj" />
```
Comment (eerhardt) [inline src/Microsoft.ML.AutoML/Microsoft.ML.AutoML.csproj]: Why aren't you using a `PackageReference` to Recommender here, same as the other 3?

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj [hunk 5]
```diff
@@ -15,98 +18,10 @@
   <ItemGroup>
     <ProjectReference Include="..\Microsoft.ML.AutoML\Microsoft.ML.AutoML.csproj" />
   </ItemGroup>
-  
-  <ItemGroup>
```
Comment (eerhardt) [inline src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: I don't think you want these to be removed, do you?

File: Microsoft.ML.AutoML.sln (hunk 3 above)
Comment (maryamariyan) [reply Microsoft.ML.AutoML.sln]: Refer to https://github.com/dotnet/machinelearning/pull/4246/files#diff-466b4b1a71c2aede1d30fa02d412ddc3

File: pkg/Microsoft.ML.AutoML/Microsoft.ML.AutoML.nupkgproj [hunk 6]
```diff
@@ -4,11 +4,11 @@
     <TargetFramework>netstandard2.0</TargetFramework>
     <PackageDescription>ML.NET AutoML: Optimizes an ML pipeline for your dataset, by automatically locating the best feature engineering, model, and hyperparameters</PackageDescription>
   </PropertyGroup>
-
   <ItemGroup>
     <PackageReference Include="Microsoft.ML" Version="1.3.1" />
     <PackageReference Include="Microsoft.ML.LightGBM" Version="1.3.1" />
```
Comment (LittleLittleCloud) [inline pkg/Microsoft.ML.AutoML/Microsoft.ML.AutoML.nupkgproj]: shouldn't it be `LightGbm` ?

File: Microsoft.ML.AutoML.sln (hunk 2 above)
Comment (maryamariyan) [reply Microsoft.ML.AutoML.sln]: no it's not. I can remove this

File: src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/TrainerGenerators.cs [hunk 7]
```diff
@@ -556,5 +556,35 @@ public override string[] GenerateUsings()
...
+            internal override IDictionary<string, string> NamedParameters
+            {
+                get
+                {
+                    return
+                    new Dictionary<string, string>()
+                    {
+                        {"MatrixColumnIndexColumnName","matrixColumnIndexColumnName" },
+                        {"MatrixRowIndexColumnName","matrixRowIndexColumnName" },
+                        {"FeatureColumnName","featureColumnName" },
```
Comment (eerhardt) [inline src/Microsoft.ML.CodeGen/CodeGenerator/CSharp/TrainerGenerators.cs]: I don't see a `FeatureColumnName` on MatrixFactorization

File: src/Microsoft.ML.AutoML/TaskKind.cs [hunk 8]
```diff
@@ -9,5 +9,6 @@ internal enum TaskKind
         BinaryClassification,
         MulticlassClassification,
         Regression,
+        Recommendation
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.AutoML/TaskKind.cs]: Please add trail comma :)

Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/TaskKind.cs]: To keep the code style in format

Comment (maryamariyan) [reply src/Microsoft.ML.AutoML/TaskKind.cs]: note: I could change but this is cherry pick though. it is without trail comma in master branch as well.

Comment (eerhardt) [reply src/Microsoft.ML.AutoML/TaskKind.cs]: This is code in `master`. We shouldn't be changing it here. If we want a change like that, make it in the `master` branch.

Honestly, the only thing this branch should be used for is Code Gen and mlnet.csproj going forward. Until they are merged into `master`. Then this branch should go away.

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj [hunk 9]
```diff
@@ -7,6 +7,9 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.CodeAnalysis" Version="2.10.0" />
+    <!--<PackageReference Include="Microsoft.ML.AutoML" Version="0.16.0-preview3-28218-2" />-->
+    <PackageReference Include="Microsoft.ML" Version="1.4.0-preview3-28218-2" />
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: Why we need `Microsoft.ML` package in `CodeGen`.

Please remove comment BTW

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj (hunk 5 above)
Comment (LittleLittleCloud) [inline src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: Why we remove tt file in this case

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 10]
```diff
@@ -273,6 +273,11 @@ internal static class Defaults
 
         PredictionKind ITrainer.PredictionKind => PredictionKind.Recommendation;
 
+        /// <summary>
+        /// The <see cref="TrainerInfo"/> contains general parameters for this trainer.
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: Seems that you just do a useless copy-paste here (see the change below). Could you just recover it back since it's unnecessary

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs [hunk 11]
```diff
@@ -2,9 +2,11 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
+using System.IO;
```
Comment (LittleLittleCloud) [inline test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs]: The import order seems not right,,, doesn't StyleCop not grumble about it?

File: test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs [hunk 12]
```diff
@@ -60,5 +62,92 @@ public void AutoFitRegressionTest()
 
             Assert.IsTrue(result.RunDetails.Max(i => i.ValidationMetrics.RSquared > 0.9));
         }
+
+        [TestMethod]
+        public void AutoFitRecommendationTest()
```
Comment (LittleLittleCloud) [inline test/Microsoft.ML.AutoML.Tests/AutoFitTests.cs]: I like this end-to-end test, seems really insightful.

File: src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj (hunk 5 above)
Comment (maryamariyan) [reply src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: bringing back

Comment (maryamariyan) [reply src/Microsoft.ML.CodeGen/Microsoft.ML.CodeGenerator.csproj]: bringing back

**Comment by LittleLittleCloud** [comment]
Could you do me a favor?
Adding this line of code to `src\Microsoft.ML.AutoML\Assembly.cs`

```
[assembly: InternalsVisibleTo("Microsoft.ML.ModelBuilder.AutoMLService, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
```

Thanks!

**Comment by maryamariyan** [comment]
Thanks for the review. Halting progress on the PR and could reopen this PR once the blockers for this work item have been identified.

File: src/Microsoft.ML.AutoML/TrainerExtensions/RecommendationTrainerExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,36 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Microsoft.ML.Trainers;
+
+namespace Microsoft.ML.AutoML
+{
+    using ITrainerEsitmator = ITrainerEstimator<IPredictionTransformer<object>, object>;
```
Comment (danmosemsft) [inline src/Microsoft.ML.AutoML/TrainerExtensions/RecommendationTrainerExtensions.cs]: typo ITrainerEsitmator
